### PR TITLE
add(unit test): add negative tests to perform resize

### PR DESCRIPTION
**Pull request by mittachaitu**
Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>
This PR adds negative test cases to test resize functionality.

### CODE REVIEW

File: tests/cstor/gtest/test_zrepl_prot.cc [hunk 1]
```diff
@@ -296,6 +296,42 @@ static void get_zvol_status(std::string zvol_name, uint64_t &ioseq, int control_
...
 }
 
+static void wait_for_replica_status(std::string zvol_name, uint64_t &ioseq,
+    int control_fd, int state, struct zrepl_status_ack &status,
+    int version = REPLICA_VERSION) {
+	zvol_io_hdr_t hdr_in, hdr_out = {0};
+	int rc, max_retry_count=10, retry_count = 0;
+
+	while (retry_count < max_retry_count) {
+		hdr_out.version = version;
```
Comment (pawanpraka1) [inline tests/cstor/gtest/test_zrepl_prot.cc:275]: this looks duplicate code. Please use get_zvol_status for getting the status. see if you can return status and reuse the same function.

File: tests/cstor/gtest/test_zrepl_prot.cc [hunk 2]
```diff
@@ -296,6 +296,42 @@ static void get_zvol_status(std::string zvol_name, uint64_t &ioseq, int control_
...
+		ASSERT_ERRNO("read", rc >= 0);
+		ASSERT_EQ(rc, sizeof (hdr_in));
+		EXPECT_EQ(hdr_in.version, hdr_out.version);
+		EXPECT_EQ(hdr_in.opcode, ZVOL_OPCODE_REPLICA_STATUS);
+		EXPECT_EQ(hdr_in.io_seq, ioseq);
+		EXPECT_EQ(hdr_in.status, ZVOL_OP_STATUS_OK);
+		EXPECT_EQ(hdr_in.len, sizeof (status));
+		rc = read(control_fd, &status, sizeof (status));
+		ASSERT_ERRNO("read", rc >= 0);
+		if (status.state == state) {
```
Comment (pawanpraka1) [inline tests/cstor/gtest/test_zrepl_prot.cc:297]: We should fail if we exceeded the max_retry_count. Here we just assume that vol is healthy if we have exceeded the max count.

Comment (pawanpraka1) [reply tests/cstor/gtest/test_zrepl_prot.cc:297]: never mind, I see you checking the status again after returning from this function.

File: tests/cstor/gtest/test_zrepl_prot.cc [hunk 3]
```diff
@@ -1972,26 +2097,10 @@ TEST(ZvolResizeTest, ResizeZvol) {
...
-	ASSERT_EQ(rc, sizeof (resize_data));
-
-	rc = read(control_fd, &hdr_in, sizeof (hdr_in));
-	ASSERT_EQ(rc, sizeof (hdr_in));
-	EXPECT_EQ(hdr_in.status, ZVOL_OP_STATUS_OK);
-	ASSERT_EQ(hdr_in.len, 0);
-
+	execute_resize_opcode(new_size, zvolname, control_fd, ZVOL_OP_STATUS_OK);
 	// get the clone zvol size after triggering resize
 	str = execCmd("zfs", std::string("get -Hpo value volsize ") +
```
Comment (pawanpraka1) [inline tests/cstor/gtest/test_zrepl_prot.cc:2082]: here check main_zv should have old size also.

File: tests/cstor/gtest/test_zrepl_prot.cc (hunk 1 above)
Comment (mittachaitu) [reply tests/cstor/gtest/test_zrepl_prot.cc:275]: I will reuse it

File: tests/cstor/gtest/test_zrepl_prot.cc (hunk 2 above)
Comment (pawanpraka1) [reply tests/cstor/gtest/test_zrepl_prot.cc:297]: its better to put that check here in this function only.

File: tests/cstor/gtest/test_zrepl_prot.cc (hunk 1 above)
Comment (mittachaitu) [reply tests/cstor/gtest/test_zrepl_prot.cc:275]: Done

File: tests/cstor/gtest/test_zrepl_prot.cc (hunk 2 above)
Comment (mittachaitu) [reply tests/cstor/gtest/test_zrepl_prot.cc:297]: Yes done

File: tests/cstor/gtest/test_zrepl_prot.cc [hunk 4]
```diff
@@ -1934,27 +1943,116 @@ TEST(ZvolResizeTest, DataConn) {
...
+	str = execCmd("zfs", std::string("get -Hpo value volsize ") +
+			zvolname);
+	main_zv_size = atoi(str.c_str());
+	EXPECT_EQ(new_size, main_zv_size);
+
+	graceful_close(control_fd);
+}
+
+/*
+ * Test zvol resize when it is degrade
```
Comment (pawanpraka1) [inline tests/cstor/gtest/test_zrepl_prot.cc]: typo degraded?

**Review by pawanpraka1 — DISMISSED** [review]
looks good.

**Review by pawanpraka1 — APPROVED** [review]
looks good.
